include/envoy/upstream/upstream.h

```diff
@@ -269,6 +273,11 @@ class HostSet {
    */
   virtual const HostVector& hosts() const PURE;
 
+  /**
+   * @return a shared ptr to the vector returned by hosts().
+   */
+  virtual HostVectorConstSharedPtr hostsPtr() const PURE;
```

Comment:
I realize this would be a huge amount of churn, but it having both `hosts()` and `hostsPtr()` in the API seems pretty odd to me since one is implemented in terms of the other. Seems like `hosts()` etc. should just return the shared_ptr? Feel free to punt for now, but maybe worth a TODO that we might consider cleaning this up at some point? WDYT?

Reply:
Yeah I agree, I just did this to avoid this PR from getting out of control. I'll add a TODO